### ISSUE TITLE
Intermittent segfault in alert.message call in alert handlers (Python binding)

### ISSUE THREAD

**Issue by cas--** [arvidn/libtorrent#6437]
I recently moved Deluge to [Github actions](https://github.com/deluge-torrent/deluge/actions) with tests running against libtorrent 1.2 branch. I have found that the test runs are [occasionally segfaulting ](https://github.com/cas--/Deluge/runs/3405529050?check_suite_focus=true) but not consistently and re-running the test usually fixes the run. I know it's not much to go on but some help with resolving it would be good as I wonder if it might affect users as well.

Prefixed the test runs with `catchsegv` to capture a backtrace:

```
2021-08-23T22:37:07.0127987Z Backtrace:
2021-08-23T22:37:07.0226902Z /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcb)[0x7f9e95f1218b]
2021-08-23T22:37:07.0303495Z /lib/x86_64-linux-gnu/libc.so.6(+0x46210)[0x7f9e95f12210]
2021-08-23T22:37:07.0383733Z /home/runner/work/Deluge/Deluge/deluge/libtorrent.cpython-38-x86_64-linux-gnu.so(+0x131cf3)[0x7f9e92760cf3]
2021-08-23T22:37:07.0463213Z /lib/x86_64-linux-gnu/libboost_python38.so.1.71.0(_ZNK5boost6python7objects8function4callEP7_objectS4_+0x2cb
```

Attached full test log [external_ip_alert_segfaultlog.zip](https://github.com/arvidn/libtorrent/files/7072543/external_ip_alert_segfaultlog.zip), see file `test-linux/7_Test with pytest.txt`. 

Although no debug symbols available I used gdb on local machine to see the last libtorrent calls:

```
$ gdb /usr/lib/python3/dist-packages/libtorrent.cpython-38-x86_64-linux-gnu.so
(gdb) x/20i 0x131cf3
   0x131cf3:	mov    -0x1(%rdx,%rax,1),%rax
   0x131cf8:	jmp    0x131ca5
   0x131cfa:	callq  0x521f0 <__stack_chk_fail@plt>
   0x131cff:	endbr64 
   0x131d03:	mov    %rax,%r12
   0x131d06:	mov    (%rsp),%rdi
   0x131d0a:	lea    0x10(%rsp),%rax
   0x131d0f:	cmp    %rax,%rdi
   0x131d12:	je     0x131d19
   0x131d14:	callq  0x52000 <_ZdlPv@plt>
   0x131d19:	mov    %r12,%rdi
   0x131d1c:	callq  0x529d0 <_Unwind_Resume@plt>
   0x131d21:	nopw   %cs:0x0(%rax,%rax,1)
   0x131d2b:	nopl   0x0(%rax,%rax,1)
   0x131d30:	endbr64 
   0x131d34:	push   %rbx
   0x131d35:	mov    %rdi,%rbx
   0x131d38:	mov    0x18(%rsi),%rdi
   0x131d3c:	mov    0xe864d(%rip),%rsi        # 0x21a390
   0x131d43:	
    callq  0x51900 <_ZN5boost6python9converter22get_lvalue_from_pythonEP7_objectRKNS1_12registrationE@plt>
```

libtorrent version (or branch): 1.2 RC branch
platform/architecture: Ubuntu 20.04

**Comment by AllSeeingEyeTolledEweSew** [comment]
Can you build with debug symbols?

I believe the following works in the RC_1_2 branch:

```shell
python setup.py build_ext --debug bdist_wheel
```

Also, have you tried `python -X dev`? It improves stack traces from python and uses a debugging memory allocator, among other things.

**Comment by AllSeeingEyeTolledEweSew** [comment]
I've seen a number of crashes that occur consistently at some test in a test suite, that are *really* triggered by `gc`, and the bug doesn't have anything to do with the crashing test. `python -X dev` helps a lot with these.

One thing I've done to convince myself (and arvid) of the cause of a bug is to write a small python script triggers the bad code path and does nothing else. If it's flaky, you can run the code path in a loop until it crashes.

**Comment by cas--** [comment]
Thanks I'll try those tips.

Just to note that these unit tests for Deluge have been running for years without any crashes like this so something has changed with 1.2 branch. e.g. the Windows tests were running fine on 1.1.13 but as soon as switched to 1.2.14 it started crashing.

Also had problems building the wheel for RC_1_2 on Ubuntu 20.04:

```
/usr/bin/ld: cannot find -ltorrent-rasterbar
collect2: error: ld returned 1 exit status

    "g++"    -o "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/libtorrent.cpython-38-x86_64-linux-gnu.so" -Wl,-h -Wl,libtorrent.cpython-38-x86_64-linux-gnu.so -shared -Wl,--start-group "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/module.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/sha1_hash.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/converters.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/create_torrent.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/fingerprint.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/utility.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/session.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/entry.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/torrent_info.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/string.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/torrent_handle.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/torrent_status.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/session_settings.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/version.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/alert.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/datetime.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/peer_info.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/ip_filter.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/magnet_uri.o" "bin/gcc-9/release/address-model-64/lt-visibility-hidden/python-3.8/src/error_code.o"  -Wl,-Bstatic  -Wl,-Bdynamic -lboost_python38 -ltorrent-rasterbar -ldl -lpthread -lutil -Wl,--end-group -fPIC -L/home/calum/projects/libtorrent/bindings/python/../../src/.libs -L/usr/lib/x86_64-linux-gnu -Wl,-Bsymbolic
```

but I did get it to build with: 

`python3 setup.py build_ext --debug --b2-args="libtorrent-link=static"  bdist_wheel`

**Comment by cas--** [comment]
I neglected to mention that the tests run fine on local machine so it might be the GitHub Actions server (network?) setup that is triggering the problem.

**Comment by cas--** [comment]
So while testing locally with libtorrent 2.0 the problem is persistent and I got the following backtrace:

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff7c0a859 in __GI_abort () at abort.c:79
#2  0x00007ffff29fa911 in  () at /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff2a0638c in  () at /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff2a063f7 in  () at /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff2a07155 in  () at /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff3000149 in boost::python::detail::invoke<boost::python::to_python_value<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (libtorrent::alert::*)() const, boost::python::arg_from_python<libtorrent::alert&> >(boost::python::detail::invoke_tag_<false, true>, boost::python::to_python_value<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (libtorrent::alert::*&)() const, boost::python::arg_from_python<libtorrent::alert&>&)
    (rc=..., f=@0x13abba8: &virtual table offset 32, tc=...) at /usr/include/boost/python/detail/invoke.hpp:86
#7  0x00007ffff2fe5d60 in boost::python::detail::caller_arity<1u>::impl<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (libtorrent::alert::*)() const, boost::python::default_call_policies, boost::mpl::vector2<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, libtorrent::alert&> >::operator()(_object*, _object*) (this=0x13abba8, args_=(<external_ip_alert() at remote 0x7fffbc4c1c50>,))
    at /usr/include/boost/python/detail/caller.hpp:233
#8  0x00007ffff2fd49af in boost::python::objects::caller_py_function_impl<boost::python::detail::caller<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (libtorrent::alert::*)() const, boost::python::default_call_policies, boost::mpl::vector2<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, libtorrent::alert&> > >::operator()(_object*, _object*)
    (this=0x13abba0, args=(<external_ip_alert() at remote 0x7fffbc4c1c50>,), kw=0x0)
    at /usr/include/boost/python/object/py_function.hpp:38
#9  0x00007ffff2b5eeab in boost::python::objects::function::call(_object*, _object*) const ()
    at /usr/lib/x86_64-linux-gnu/libboost_python38.so.1.71.0
#10 0x00007ffff2b5f15c in  () at /usr/lib/x86_64-linux-gnu/libboost_python38.so.1.71.0
#11 0x00007ffff2b64233 in boost::python::handle_exception_impl(boost::function0<void>) ()
    at /usr/lib/x86_64-linux-gnu/libboost_python38.so.1.71.0
```

**Comment by arvidn** [comment]
@cas-- is there any output printed to `stderr`? it looks like you're hitting an `assert()` in boost-python

**Comment by arvidn** [comment]
is there any risk that the alert objects are saved and then accessed *after* another call to `pop_alerts()`?

whenever `pop_alerts()` is called, all alert objects returned by previous calls are invalidated. That could possibly explain that the `std::string` being returned is invalid (i.e. is a dangling pointer).

Another possibility is that some parts of the program are built with incompatible configurations. Does it still happen if libtorrent and boost are built and linked statically against the python binding module?

**Comment by cas--** [comment]
It is possible that the object is accessed after another call to pop_alerts :thinking: 

I have been running our test suite for years against libtorrent without segfaults so something has changed from lt 1.2 onwards and seems quite specific to external_ip alert. 

For reference this is the alert manager and handler: 

https://github.com/deluge-torrent/deluge/blob/develop/deluge/core/alertmanager.py#L108
https://github.com/deluge-torrent/deluge/blob/develop/deluge/core/torrentmanager.py#L1638

**Comment by AllSeeingEyeTolledEweSew** [comment]
https://github.com/deluge-torrent/deluge/blob/1e6cc0394666ae95e9521f1016d07d56a3ed43a5/deluge/core/alertmanager.py#L137-L143

If the intent is to copy all the data from an alert before it gets invalidated by the next `pop_alerts()`, I'm dubious that this is sufficient.

For example, I think `alert_copy.message` just points to `alert.message`, which is a bound method. So `alert_copy.message()` will refer to alert memory that might've been invalidated.

For this to not crash, there would have to be a guarantee that the callbacks registered by `recator.callLater()` always complete before the next `update()`, which seems doubtful.

Do you have debug logs that happen before a crash? You should be able to confirm whether you dispatch a callback for `external_ip_alert`, but `pop_alerts()` before the callback runs.

**Comment by cas--** [comment]
Ha so the copying of alerts was to fix sefaults in lt 1.2 alerts! As mentioned in the my commit https://github.com/deluge-torrent/deluge/commit/be74d96c6a665b18284d2a470eefdd1e7cde8527 that closed #2779 

Thanks for patience and pointers, I understand the issue better now, it's been so long since I touched that code :roll_eyes:

**Comment by AllSeeingEyeTolledEweSew** [comment]
Ahh, alert memory worked differently in 1.1? That would explain this.

See also #5154 

@arvidn could you reopen #5154 actually? it got railroaded by stale-bot

**Comment by arvidn** [comment]
yeah, I'm sorry about the more complicated handling of alerts since libtorrent 1.1, especially since it's hard to map well to python.  however, I still think it was worth [it](https://blog.libtorrent.org/2015/04/libtorrent-alert-queue/).

**Comment by AllSeeingEyeTolledEweSew** [comment]
@cas-- check out https://github.com/AllSeeingEyeTolledEweSew/tvaf/blob/master/tvaf/driver.py for my approach to alert memory.

TL;DR instead of using alert-handling callbacks, I created iterators for alerts. This is possible if you coordinate `pop_alerts()` to happen while every iterator is blocked waiting for their next alert.

I had trouble with the alert-handling callback architecture. Let's say you have a narrowly-scoped task: your app serves data, and you want to call `read_piece()` and waiting for the corresponding `read_piece_alert`, and send the data somewhere. You have a few choices:
* Register an ad-hoc `on_read_piece_alert` handler function, which sends the data to the right place. This works fine, but consider if the app wants to cancel the read: you can unregister the handler, but it may have already been scheduled as a callback, so it may run after you unregister it even in a single-threaded environment. The handler needs to reference the invoking read job, and avoid sending data if the read was cancelled. If the handler spawns its *own* callbacks, *they* must be aware of cancellation as well. You could write some generic cancellation tokens to help with this, but this gets verbose.
* Have a few alert-handling callbacks which last most or all the lifecycle of the app. The handlers are aware of all possible tasks. The tasks are generally stored as data objects rather than coroutines, functions or code objects.

I think deluge mostly uses the second pattern.

Using alert iterators allows tasks to use normal control flow. I found it hugely simplifies development.

**Comment by cas--** [comment]
I have found a quick fix by replacing alert.message() usage in the external_ip handler. 

@AllSeeingEyeTolledEweSew Thanks for the detailed options!

@arvidn Is there a cleaner way to handle this pointer issue in Boost Python so the app doesn't hard crash? e.g. generate RuntimeError instead

**Comment by arvidn** [comment]
it would be possible to wrap the alert pointers, in the python binding, with objects that also track whether the underlying pointer is still valid. That way it could turn into a run-time error instead. It's a fair amount of work, and I don't know what the best way would be to notify those wrappers of the pointers being invalidated.

Another approach, suggested by @AllSeeingEyeTolledEweSew is to wrap the whole container/list of alert pointers and be able to *detach* them from the session. That wouldn't in itself solve the problems, but it would allow for simpler use of the alerts (most likely).

**Comment by cas--** [comment]
Thanks I thought I would ask if there was a solution to keep the application alive. 

So after fixing the issue in external_ip_alert handler, the issue has progressed to other handlers that call `alert.message()` but seemingly not other preceding calls such as `alert.error.message()`. Is there something specific with alert.message() that means it is more likely to be invalidated?
